### PR TITLE
release.yml

### PR DESCRIPTION
**Pull request by dHannasch**
The original plan for https://github.com/pypa/packaging/issues/339 was to use `workflow_dispatch` to manually trigger the release, but it looks like in the meantime it's become possible to hook onto creating a release on https://github.com/pypa/packaging/releases, so I'm guessing we'll want to do that.

That raises an interesting question, though. If the process is initiated from https://github.com/pypa/packaging/releases, then the tag will be applied to the current head of the branch, meaning no matter what shenanigans we pull in the actions, `_bump` cannot change the `__version__` on the commit that will get tagged...because that commit has already been selected, on https://github.com/pypa/packaging/releases.
We can do what https://github.com/grst/python-ci-versioneer does and rewrite the `__version__` on the fly when publishing, but that still requires a manual call of `_bump` to update the version in the source repository (whether through `nox` or any other way). That *could* be done from GitHub Actions too, though that would mean GitHub Actions simultaneously making-a-change-and-pushing-to-PyPI and making-the-change-and-pushing-to-GitHub.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  deploy:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.x'
```
Comment (pradyunsg) [inline .github/workflows/release.yml:20]: nit: This is redundant

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        PLACEHOLDER='__version__ = "20.10.dev0"'
+        VERSION_FILE='packaging/__about__.py'
+        # Ensure the placeholder is there. If grep doesn't find the placeholder,
+        # it exits with exit code 1 and GitHub Actions aborts the build.
+        grep "$PLACEHOLDER" "$VERSION_FILE"
+        sed -i "s/$PLACEHOLDER/__version__ = \"${VERSION}\"/g" "$VERSION_FILE"
+        # Optionally, we could git push here, similar to what the nox session does,
+        # though that means git push from GitHub Actions.
+        # https://github.com/ad-m/github-push-action
+      shell: bash
```
Comment (pradyunsg) [inline .github/workflows/release.yml:45]: This means that we'd be publishing source doe that's different from what was tagged.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    - name: Publish to test.pypi.org
+      uses: pypa/gh-action-pypi-publish@master
+      with:
+        password: ${{ secrets.TEST_PYPI_API_TOKEN }}
+        repository_url: https://test.pypi.org/legacy/
+    - name: Publish package
+      uses: pypa/gh-action-pypi-publish@master
+      with:
+        user: __token__
+        password: ${{ secrets.PYPI_API_TOKEN }}
```
Comment (pradyunsg) [inline .github/workflows/release.yml:55]: Let's not upload to PyPI right now, and instead just publish a release to TestPyPI for these initial iterations.

**Review by pradyunsg — CHANGES_REQUESTED** [review]
Hmm... I'm not a fan of this approach for automating the release process. We've got a bunch of bash-foo in a YAML file and we're modifying the tagged files before making a release.

**Review by glochlainn — APPROVED** [review]

File: .github/workflows/release.yml (hunk 2 above)
Comment (di) [reply .github/workflows/release.yml:45]: +1, I don't think this is a good pattern.

**Comment by brettcannon** [comment]
Since changes were requested but not made over a year ago, I'm closing this PR as stale.

**Pull request closed without merging**